cmd/gen-hostcert/main.go

```diff
@@ -0,0 +1,203 @@
+// Copyright 2023 Yahoo Inc.
+// Licensed under the terms of the Apache License 2.0. Please see LICENSE file in project root for terms.
+
+package main
+
+import (
+	"context"
+	"crypto"
+	"crypto/x509"
+	"encoding/json"
+	"encoding/pem"
+	"flag"
+	"fmt"
+	"log"
+	"net"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/theparanoids/crypki"
+	"github.com/theparanoids/crypki/config"
+	"github.com/theparanoids/crypki/pkcs11"
+	"github.com/theparanoids/crypki/proto"
+	"github.com/theparanoids/crypki/server/scheduler"
+	"github.com/theparanoids/ysshra/csr/transid"
+	"github.com/theparanoids/ysshra/keyid"
+	"github.com/theparanoids/ysshra/sshutils/key"
+
+	"golang.org/x/crypto/ssh"
+)
+
+const (
+	hostPrivKeyPath = "./hostkey"
+	hostPubKeyPath  = "./hostkey.pub"
+	hostCertPath    = "./hostkey-cert.pub"
+)
+
+var (
+	cfg          string
+	principals   []string
+	validityDays uint64
+	keyType      string
+	reqUser      string
+	keyAlgorithm key.PublicKeyAlgo
+)
+
+func parseFlags() {
+	var prins string
+	flag.StringVar(&cfg, "config", "", "CA key configuration file")
+	flag.Uint64Var(&validityDays, "days", 730, "validity period in days")
+	flag.StringVar(&prins, "prins", "", "principal list")
+	flag.StringVar(&keyType, "keyType", "", "key algorithm type to generate key/cert pair (supported options: RSA2048, ECCP256, ECCP384)")
+	flag.StringVar(&reqUser, "reqUser", "devops", "user to request the certificate")
+	flag.Parse()
+	log.SetFlags(log.LstdFlags | log.Lshortfile)
+
+	if cfg == "" {
+		log.Fatal("no CA key configuration file specified")
+	}
+
+	var err error
+	keyAlgorithm, err = key.GetSSHKeyAlgo(keyType)
+	if err != nil {
+		log.Printf("warning: %v", err)
+	}
+
+	principals = strings.Split(prins, ",")
+}
+
+func constructUnsignedSSHCert(pub ssh.PublicKey) (*ssh.Certificate, error) {
+	hostname, err := os.Hostname()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get hostname: %v", err)
+	}
+
+	keyID := keyid.KeyID{
+		Version:       1,
+		Principals:    principals,
+		TransID:       transid.Generate(),
+		TouchPolicy:   1,
+		IsHWKey:       false,
+		IsHeadless:    true,
+		IsFirefighter: false,
+		ReqUser:       reqUser,
+		ReqHost:       hostname,
```

Comment:
can we allow specifying hostname and localIP(s?) as a command-line flag? If none is specified, we can use `os.Hostname()` and `getLocalIP()` respectively.